Random key generation

piotrpalcz opened the pull request:
This PR is realted to Shim/rune PR https://github.com/confidential-containers/enclave-cc/pull/256 and enclave-cc PR https://github.com/confidential-containers/enclave-cc/pull/265

Creates a randomly generated key to use in unionFS that is then stored in file and passed to be mounted and read by boot enclave to use that unionFS

arronwy commented:
As described in: https://rust-random.github.io/book/guide-rngs.html#not-a-crypto-library
    "It should be emphasised that this is not a cryptography library"
    "It is therefore recommended to use specialized libraries where possible,
    for example openssl, ring and the RustCrypto libraries."

`image-rs` also have two features `encryption-openssl` and ` encryption-ring`  for them.

piotrpalcz commented:
> As described in: https://rust-random.github.io/book/guide-rngs.html#not-a-crypto-library "It should be emphasised that this is not a cryptography library" "It is therefore recommended to use specialized libraries where possible, for example openssl, ring and the RustCrypto libraries."
> 
> `image-rs` also have two features `encryption-openssl` and ` encryption-ring` for them.

Updated, but I need to explain :
In issues regarding this matter ( https://github.com/confidential-containers/enclave-cc/issues/20 & https://github.com/confidential-containers/enclave-cc/issues/149) there wasnt specified that this key needs to be cryptographically secure and generated with specific library. I just chose the easiest approach at the time.

Now I updated the branch using already existing in Ocicrypt-rs module rand_bytes function, which does exactly what I needed here and is already being drived by `cfg feature` with default value in its cfg. This though needed me to change it's access level from pub(crate) to pub.

piotrpalcz commented:
Closing, replacing with https://github.com/confidential-containers/enclave-cc/pull/312

mythi commented on ocicrypt-rs/src/blockcipher/mod.rs:
```diff
@@ -12,7 +12,7 @@ use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
 use aes_ctr::AESCTRBlockCipher;
 
-mod rand;
+pub mod rand;
```
we are getting ready with the work. @arronwy can you comment on this part. looks OK?

arronwy replied:
yes, it's ok from my side.

mythi commented on image-rs/src/snapshots/occlum/unionfs.rs:
```diff
@@ -124,18 +143,48 @@ impl Snapshotter for Unionfs {
+        let random_key = generate_random_key();
+        let new_key_path = Path::new("/new_key");
+        fs::create_dir_all(new_key_path)?;
+        std::fs::write(&PathBuf::from(&new_key_path.join("key.txt")), &random_key)?;
```
could this be:
```suggestion
        std::fs::write(&PathBuf::from(&keys_mount_path.join("key.txt")), &random_key)?;
```
and moved to the end to replace `fs::copy()`

mythi reviewed (approved):
LGTM and we have enclave-cc CI passing with this

mythi commented on image-rs/src/image.rs:
```diff
@@ -526,7 +526,7 @@ mod tests {
             "mcr.microsoft.com/hello-world",
             // Docker container Registry
-            "docker.io/i386/busybox",
+            "docker.io/busybox",
```
This is from my earlier commit

piotrpalcz replied:
Its cherry picked, so you remain author. Had some github with that particular rebase but now its all fine.

piotrpalcz replied:
Fixed git history now

mythi commented:
@arronwy does this look OK?

arronwy commented on image-rs/src/snapshots/occlum/unionfs.rs:
```diff
@@ -124,18 +141,41 @@ impl Snapshotter for Unionfs {
+
         let options = format!(
             "dir={},key={}",
             Path::new("/images").join(cid).join("sefs/lower").display(),
```
we can also use `key_mount_options` here.·

arronwy reviewed (approved):
Thanks @piotrpalcz LGTM! Just a minor suggest.
